CIM-seq
================

Release build: <a href="https://travis-ci.com/EngeLab/CIMseq"><img src="https://travis-ci.com/EngeLab/CIMseq.svg?branch=master"></a>

Devel build: <a href="https://travis-ci.com/EngeLab/CIMseq"><img src="https://travis-ci.com/EngeLab/CIMseq.svg?branch=devel"></a>

Test coverage release: [![Coverage status](https://codecov.io/gh/EngeLab/CIMseq/branch/master/graph/badge.svg)](https://codecov.io/github/EngeLab/CIMseq?branch=master)

Test coverage devel: [![Coverage status](https://codecov.io/gh/EngeLab/CIMseq/branch/devel/graph/badge.svg)](https://codecov.io/github/EngeLab/CIMseq?branch=devel)

Advances in single-cell biology has enabled us to investigate isolated single cells at an unprecedented scale and resolution. However, cells in multicellular organisms are largely defined by their spatial organization within organ structures, which means that general methods for studying direct cell interaction on a large scale are needed. We propose a novel method, Cell Interaction by Multiplet Sequencing (CIM-Seq) that uses computational deconvolution of RNA-seq data from partially dissociated tissue to create cell interaction maps. We applied CIM-seq to human fetal pancreas, demonstrating that it recapitulates known cell interactions such as acinar-ductal cell contacts. Furthermore, we discover a strong link between a mesenchymal cell subtype and endocrine progenitor cells, and identify the set of genes that distinguishes this mesenchymal subtype. Thus, CIM-Seq is a general method for cell interaction studies that can be used on cell types defined to an arbitrary resolution allowing identification of interacting sub-cell types or cell states.
