CIM-seq
================

Release build:
<a href="https://travis-ci.org/jasonserviss/sp.scRNAseq"><img src="https://travis-ci.com/jasonserviss/sp.scRNAseq.svg?token=8VXNvJpmRV9RpFtTcJYW&branch=master"></a>

Devel build:
<a href="https://travis-ci.org/jasonserviss/sp.scRNAseq"><img src="https://travis-ci.com/jasonserviss/sp.scRNAseq.svg?token=8VXNvJpmRV9RpFtTcJYW&branch=devel"></a>

Test coverage: [![Coverage
status](https://codecov.io/gh/jasonserviss/sp.scRNAseq/branch/master/graph/badge.svg?token=vVxKKSzhGt)](https://codecov.io/gh/jasonserviss/sp.scRNAseq)

Advances in single-cell biology has enabled us to investigate isolated
single cells at an unprecedented scale and resolution. However, cells in
multicellular organisms are largely defined by their spatial
organization within organ structures, which means that general methods
for studying direct cell interaction on a large scale are needed. We
propose a novel method, Cell Interaction by Multiplet Sequencing
(CIM-Seq) that uses computational deconvolution of RNA-seq data from
partially dissociated tissue to create cell interaction maps. We applied
CIM-seq to human fetal pancreas, demonstrating that it recapitulates
known cell interactions such as acinar-ductal cell contacts.
Furthermore, we discover a strong link between a mesenchymal cell
subtype and endocrine progenitor cells, and identify the set of genes
that distinguishes this mesenchymal subtype. Thus, CIM-Seq is a general
method for cell interaction studies that can be used on cell types
defined to an arbitrary resolution allowing identification of
interacting sub-cell types or cell states.
