---
title: "sp.scRNAseq Vignette"
author: "Jason T. Serviss"
date: '`r Sys.Date()`'
output:
    html_document:
        theme: flatly
        toc: yes
        code_folding: hide
vignette: >
    %\VignetteIndexEntry{ClusterSignificance Vignette} 
    %\VignetteEngine{knitr::rmarkdown} 
    %\usepackage[utf8]{inputenc}
    %\SweaveUTF8
---

<style>
    .col2 {
        columns: 2 200px;         /* number of columns and width in pixels*/
        -webkit-columns: 2 200px; /* chrome, safari */
        -moz-columns: 2 200px;    /* firefox */
    }
</style>

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

##Introduction
Interaction maps from incompletely dissociated primary tissue
Advances in singlecell
biology has enabled us to investigate isolated single cells at an
unprecedented resolution. Current methods can measure subtle changes in cell state, revealing
specialized minor subpopulations within a cell type. However, functional characterization of such
subpopulations is still challenging. A promising approach to untangling functional
characterization is spatial transcriptomics, where genomewide
measurements of gene
expression are enriched with spatial attributes. Such approaches allow us to determine the
relative positions of cell subpopulations. Unfortunately, current spatial transcriptomic
approaches suffer either from problems with sensitivity, are limited to a restricted panel of
genes, or the spatial attributes are of low resolution.
In order to deconvolute the cell connectome, we sort multiplets of incompletely dissociated cells
along with single cells from the same sample. The single cells provide a set of blueprints of
possible subcell
types, and subsequently we machine learning approaches to infer the
composition of multiplets. Using this method we can measure the composition of a very large
number of multiplets onto a lattice consisting of subcell
types of arbitrary complexity.

## ERCC counts 
```{r erccPlot, fig.align='center', fig.width=12, fig.height=10}
library(sp.scRNAseq)
data(expData)
spPlot(expData, type="ercc")
```

## Markers {.tabset}
### INS and GCG
```{r marker1, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('INS', 'GCG'))
```

No INS/GCG doublets

### INS and THY1
```{r marker2, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('INS', 'THY1'))
```

Only doublets, should be bona fide... These might be an interesting subpopulation to check out

### GCG and THY1
```{r marker3, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('GCG', 'THY1'))
```

Also interesting

### INS and NEUROG3
```{r marker4, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('INS', 'NEUROG3'))
```

Frequently coexpressed in single cells

### NEUROG3 and THY1
```{r marker5, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('NEUROG3', 'THY1'))
```

One doublet neurogenin+ / mesenchymal cells.

### FLT1 and THY1
```{r marker6, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('FLT1', 'THY1'))
```

Endothelial / Mesenchymal doublets

### EPCAM and THY1
```{r marker7, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('EPCAM', 'THY1'))
```

Epithelial / Mesenchymal doublets

### PROM1 and THY1
```{r marker8, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('PROM1', 'THY1'))
```

Ductal / Mesenchymal doublets

### PROM1 and FLT1
```{r marker9, fig.align='center', fig.width=12, fig.height=10}
spPlot(expData, type="markers", markers=c('PROM1', 'FLT1'))
```

Ductal / Endothelial doublets





<br></br><br></br>