---
title: "Plotting"
author: "Jason T. Serviss"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, message = FALSE}
packages <- c(
    "sp.scRNAseq",
    "sp.scRNAseqData",
    "printr",
    "ggthemes",
    "tidyverse",
    "viridis"
)
purrr::walk(packages, library, character.only = TRUE)
rm(packages)
```

```{r}
sng <- stringr::str_detect(colnames(countsMgfp), "^s")
cObjSng <- spCounts(countsMgfp[, sng], countsMgfpERCC[, sng])
cObjMul <- spCounts(countsMgfp[, !sng], countsMgfpERCC[, !sng])
uObj <- spUnsupervised(cObjSng, type = "max", max = 2000)
```

## spCounts

Get the plotting data so you can make a custom plot.
```{r}
plotCountsData(cObjSng, cObjMul) %>%
  head()
plotCountsData(cObjSng, cObjMul, markers = c("Lgr5", "Alpi")) %>%
  head()
```

Plot markers.
```{r}
plotCounts(cObjSng, cObjMul, type = "markers", markers = c("Lgr5", "Alpi")) +
  geom_point()
```

Plot markers final.
```{r}
plotCounts(cObjSng, cObjMul, type = "markers", markers = c("Lgr5", "Alpi")) +
  geom_point() +
  scale_color_ptol() +
  theme_few() +
  theme(legend.position = "top") +
  guides(colour = guide_legend(override.aes = list(size = 3)))
```

Plot cell number.
```{r}
plotCounts(cObjSng, cObjMul) +
  geom_jitter()

plotCounts(cObjSng, cObjMul) +
  geom_boxplot()
```

Add % ERCC on right axis.
```{r}
plotCounts(cObjSng, cObjMul) +
  geom_jitter() +
  scale_y_continuous(
    expand = c(0, 0),
    sec.axis = sec_axis(trans = ~ convertToERCC(., cObjSng, cObjMul), name = "% ERCC")
  )
```

Plot ERCC final
```{r}
plotCounts(cObjSng, cObjMul) +
  geom_jitter() +
  scale_y_continuous(
    expand = c(0, 0),
    breaks = 0:7,
    sec.axis = sec_axis(trans = ~ convertToERCC(., cObjSng, cObjMul), name = "% ERCC")
  ) +
  theme_few()
```

## spUnsupervised

Get the plotting data so you can make a custom plot.

```{r}
plotUnsupervisedData(uObj, cObjSng) %>% head()
plotUnsupervisedData(uObj, cObjSng, markers = c("Lgr5", "Alpi")) %>% head()
```

Plot the clustering results.
```{r}
p <- plotUnsupervised(uObj, cObjSng)
p + geom_point()
```

Plot the classifications.
```{r}
p + geom_point(aes(colour = Classification))
```

Plot classification with uncertainty.
```{r}
p + geom_point(aes(colour = Classification, size = Uncertainty), alpha = 0.75)
```

Plot gene expression.
```{r}
p <- plotUnsupervised(uObj, cObjSng, markers = c("Alpi", "Lgr5"))
p + geom_point(aes(colour = Alpi)) + scale_colour_viridis()
p + geom_point(aes(colour = Lgr5)) + scale_colour_viridis()
```

Plot mean of multiple markers for the same cell type.
```{r}
#use the data instead
plotUnsupervisedData(uObj, cObjSng, markers = c("Msi1", "Lgr5")) %>%
  gather(gene, value, -(Sample:Uncertainty)) %>%
  group_by(Sample) %>%
  mutate(`mean(Msi1, Lgr5)` = mean(value)) %>%
  select(`t-SNE dim 1`, `t-SNE dim 2`, `mean(Msi1, Lgr5)`) %>%
  distinct() %>%
  ggplot() +
  geom_point(aes(`t-SNE dim 1`, `t-SNE dim 2`, colour = `mean(Msi1, Lgr5)`)) +
  scale_colour_viridis()
```

Plot with multiple markers for different cell types.
```{r}
markers <- c("Alpi", "Lgr5", "Muc2", "Lyz1")
pal <- RColorBrewer::brewer.pal(length(markers), "Set1")
names(pal) <- markers
pal <- pal[order(names(pal))]

p <- plotUnsupervised(uObj, cObjSng, markers = markers, pal = pal)
p + geom_point(aes(colour = Colour)) + 
  scale_colour_identity(
    "Markers", 
    labels = names(pal),
    breaks = pal, 
    guide = "legend", 
    drop = FALSE
  )
```

Final marker plot.
```{r}
p + geom_point(aes(x = `t-SNE dim 1`, y = `t-SNE dim 2`), shape = 21, colour = "black") +
  geom_point(aes(colour = Colour)) + 
  scale_colour_identity(
    "Markers", 
    labels = names(pal),
    breaks = pal, 
    guide = "legend", 
    drop = FALSE
  ) +
  theme_few() +
  theme(legend.position = "top") +
  guides(colour = guide_legend(override.aes = list(size = 3)))
```


## spSwarm
